ext/standard/hrtime.c

```diff
@@ -0,0 +1,206 @@
+   +----------------------------------------------------------------------+
+ */
+
+/* $Id$ */
+
+#include "php.h"
+#include "zend_exceptions.h"
+#include "hrtime.h"
+
+/* {{{ */
+/* This file reuses code parts from the cross-platform timer library
+	Public Domain - 2011 Mattias Jansson / Rampant Pixels */
+
+#if PHP_HRTIME_PLATFORM_POSIX
+
+# include <unistd.h>
+# include <time.h>
+# include <string.h>
+
+#elif PHP_HRTIME_PLATFORM_WINDOWS
+
+# define WIN32_LEAN_AND_MEAN
+# include <windows.h>
+
+static uint64_t _timer_freq = 0;
+static double _timer_int = .0;
+
+#elif PHP_HRTIME_PLATFORM_APPLE
+
+# include <mach/mach_time.h>
+# include <string.h>
+static mach_timebase_info_data_t _timerlib_info;
+
+#elif PHP_HRTIME_PLATFORM_HPUX
+
+# include <sys/time.h>
+
+#elif PHP_HRTIME_PLATFORM_AIX
+
+# include <sys/time.h>
+# include <sys/systemcfg.h>
+
+#endif
+
+#define NANO_IN_SEC 1000000000
+/* }}} */
+
+static int _timer_init()
+{/*{{{*/
+#if PHP_HRTIME_PLATFORM_WINDOWS
+
+	uint64_t unused;
+	if (!QueryPerformanceFrequency((LARGE_INTEGER*) &_timer_freq) ||
+		!QueryPerformanceCounter((LARGE_INTEGER*) &unused)) {
+		return -1;
+	}
+	_timer_int = 1.0 / _timer_freq;
+
+#elif PHP_HRTIME_PLATFORM_APPLE
+
+	if (mach_timebase_info(&_timerlib_info)) {
+		return -1;
+	}
+
+#elif PHP_HRTIME_PLATFORM_POSIX
+
+	struct timespec ts = { .tv_sec = 0, .tv_nsec = 0 };
+
+	if (clock_gettime(CLOCK_MONOTONIC, &ts)) {
+		return -1;
+	}
+
+#elif PHP_HRTIME_PLATFORM_HPUX
+
+	/* pass */
+
+#elif PHP_HRTIME_PLATFORM_AIX
+
+	/* pass */
+
+#else
+	/* Timer unavailable. */
+	return -1;
+#endif
+
+	return 0;
+}/*}}}*/
+
+/* {{{ */
+PHP_MINIT_FUNCTION(hrtime)
+{
+	if (0 > _timer_init()) {
+		php_error_docref(NULL, E_WARNING, "Failed to initialize high-resolution timer");
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+/* }}} */
+
+static zend_always_inline php_hrtime_t _timer_current(void)
+{/*{{{*/
+#if PHP_HRTIME_PLATFORM_WINDOWS
+	uint64_t cur;
+	QueryPerformanceCounter((LARGE_INTEGER*) &cur);
+	return (php_hrtime_t)((double)cur * _timer_int * (double)NANO_IN_SEC);
+#elif PHP_HRTIME_PLATFORM_APPLE
+	return (php_hrtime_t)mach_absolute_time() * _timerlib_info.numer / _timerlib_info.denom;
+#elif PHP_HRTIME_PLATFORM_POSIX
+	struct timespec ts = { .tv_sec = 0, .tv_nsec = 0 };
+	if (0 == clock_gettime(CLOCK_MONOTONIC, &ts)) {
+		return ((php_hrtime_t) ts.tv_sec * (php_hrtime_t)NANO_IN_SEC) + ts.tv_nsec;
+	}
+	return 0;
+#elif PHP_HRTIME_PLATFORM_HPUX
+	return (php_hrtime_t) gethrtime();
+#elif  PHP_HRTIME_PLATFORM_AIX
+	timebasestruct_t t;
+	read_wall_time(&t, TIMEBASE_SZ);
+	time_base_to_time(&t, TIMEBASE_SZ);
+	return (php_hrtime_t) t.tb_high * (php_hrtime_t)NANO_IN_SEC + t.tb_low;
+#else
+	return 0;
+#endif
+}/*}}}*/
+
+#if ZEND_ENABLE_ZVAL_LONG64
+#define PHP_RETURN_HRTIME(t) RETURN_LONG((zend_long)t)
+#else
+#ifdef _WIN32
+# define HRTIME_U64A(i, s, len) _ui64toa_s(i, s, len, 10)
+#else
+# define HRTIME_U64A(i, s, len) \
```

Comment:
Don't we already have ZEND_LTOA? Why duplicate it?

Reply:
ZEND_LTOA depends on the platform and matches zend_long. The timer is saved in uint64_t, thus the macro is not suitable on 32-bit.